homeassistant/components/bluecurrent/__init__.py

```diff
@@ -0,0 +1,244 @@
+
+from datetime import datetime
+from typing import Any
+
+from bluecurrent_api import Client
+from bluecurrent_api.exceptions import (
+    BlueCurrentException,
+    InvalidApiToken,
+    RequestLimitReached,
+    WebsocketException,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_TOKEN, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers import entity_registry
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import DeviceInfo, Entity
+from homeassistant.helpers.event import async_call_later
+
+from .const import (
+    CHARGE_POINTS,
+    DATA,
+    DELAY_1,
+    DELAY_2,
+    DOMAIN,
+    EVSE_ID,
+    GRID_STATUS,
+    LOGGER,
+    MODEL_TYPE,
+    OBJECT,
+    PLATFORMS,
+    VALUE_TYPES,
+)
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up Blue Current as a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    client = Client()
+    api_token = config_entry.data[CONF_API_TOKEN]
+    connector = Connector(hass, config_entry, client)
+    try:
+        await connector.connect(api_token)
+    except (BlueCurrentException) as err:
+        if isinstance(err, InvalidApiToken):
+            raise ConfigEntryAuthFailed("Invalid api token.") from err
+        raise ConfigEntryNotReady from err
+
+    hass.loop.create_task(connector.start_loop())
+    await client.get_charge_points()
+
+    await client.wait_for_response()
+    hass.data[DOMAIN][config_entry.entry_id] = connector
+    hass.config_entries.async_setup_platforms(config_entry, PLATFORMS)
+
+    async def _async_disconnect_websocket(_: Event) -> None:
+        await connector.disconnect()
+
+    config_entry.async_on_unload(
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, _async_disconnect_websocket
+        )
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload the BlueCurrent config entry."""
+    connector: Connector = hass.data[DOMAIN].pop(config_entry.entry_id)
+    hass.async_create_task(connector.disconnect())
+
+    return await hass.config_entries.async_unload_platforms(config_entry, PLATFORMS)
+
+
+def set_entities_unavalible(hass: HomeAssistant, config_id: str) -> None:
+    """Set all Blue Current entities to unavailable."""
+    registry = entity_registry.async_get(hass)
+    entries = entity_registry.async_entries_for_config_entry(registry, config_id)
+
+    for entry in entries:
+        entry.write_unavailable_state(hass)
+
+
+class Connector:
+    """Define a class that connects to the Blue Current websocket API."""
+
+    charge_points: dict[str, dict] = {}
+    grid: dict[str, Any] = {}
```

Comment:
Unless you mean these to be shared between `Connector` instances, don't define them here – you can leave them as type hints, but assign them in the constructor:

```suggestion
    charge_points: dict[str, dict]
    grid: dict[str, Any]
```

Reply:
Fixed.